### ISSUE TITLE
Automate enforcement of deprecated parameter removal dates

### ISSUE THREAD

**Issue by sapitts** [idaholab/moose#18631]
## Reason
At present the determination of the time interval until removal of deprecated parameters is left to the developers. Often these deprecated parameters are not promptly removed on the specified deprecated date. Inconsistencies in the formatting of the date add to the difficulties in reviewing and identifying the deprecation dates.

## Design
Develop an automated method to determine if the deprecated date has past and trigger some notifying mechanism, e.g. warning, error, refusal to build code, etc. Additionally develop a standardized method to set the time interval for a deprecated parameter to be allowed to stay in the code (6 months?)
Tagging @dschwen because these are all his ideas; I'm just recording the thoughts from the standup meeting

## Impact
Better alignment of given dates for the removal of deprecated parameters and the actual removal from the code base

**Comment by dschwen** [comment]
As discussed in the standup today, I'd suggest making deprecated functionality fail hard, but offer a command line option to downgrade those errors to warnings (for people who do not want or cannot update their inputs for whatever reason, e.g. an imminent deadline etc.)

**Comment by dschwen** [comment]
PIng @moose-ccb , this seems like a broader interest framework level thing

**Comment by hugary1995** [comment]
This issue might be related to #17312

**Comment by sapitts** [comment]
@hugary1995 yes, the issues are related. I hadn't intended to create a duplicated issue but did do anyway. @dschwen would you close your least favorite of these two issues please?

**Comment by rwcarlsen** [comment]
The concern is that a hard-error approach causes tests to fail based on the current date/time - i.e. a simulation that ran yesterday won't run today even though the binary, inputs, etc. didn't change at all.  This sort of behavior is surprising, confusing, and can often result in moose not running for someone who is powerless to fix the problem.  Users need to be able to "revert" their binary/environment to a state that will allow them to get stuff done.  If we trigger based on date, they can't do that - because moose just started failing despite their environment not changing.  I think we need to keep deprecated->error transitions to only occuring with app version/binary updates.

**Comment by fdkong** [comment]
Greed with @rwcarlsen 

Based on date will be super painful. Release reversion will make a lot more sense. That is what PETSc is doing.

**Comment by aeslaughter** [comment]
We had this ability in the past. A warning changing to an error caused more problems then it was worth.

**Comment by roystgnr** [comment]
I'm a fan of the libMesh '--disable-deprecated' configure flag.  Users who want to be as sure as possible that their old code still works unchanged leave it off.  Users who want to be as sure as possible that their current code will be supported in the future turn it on and fix whatever it screams about.  Users who want their old unchanged code to be supported in the future are still out of luck, but there's no way to fix that without just never making backwards-incompatible API changes.  If we're going to force people to upgrade the nicest thing we can do is to let them have finer control over when.

With libMesh though there's just one handwavy deprecated level, "we plan on getting rid of this someday", vs "we don't have any *current* plans to get rid of this..."  It wouldn't be crazy to let people set e.g. "--with-deprecated-date=20220601" and throw errors for deprecated code if-and-only-if its support is intended to be removed by 2022-06-01.

**Comment by YaqiWang** [comment]
I always hope deprecation capability is not for the laziness of developers to update all the input files in the repositories. So theoretically speaking, we can always push in a patch for Civet to maintain and Civet will apply this patch at the deadline. I do not quite like the idea of keeping the deprecated code around even after the supposed last date.

**Comment by dschwen** [comment]
@rwcarlsen I agree with most of this, except the "powerless to fix the problem" part. That's exactly what my top comment addresses, a CMD line switch to turn those errors back into warnings to permit the user to run stuff. The deprecation error could mention this possiblilty explicitly.

I do definitely like attaching the deprecation point to releases, _but_ that would require a _release schedule_, otherwise it is just a wild guess what a deprecation point actually means. Will the functionality be gone in a year, next month, tomorrow?

**Comment by rwcarlsen** [comment]
I like the idea of a command line flag, but I'd prefer that by default things don't auto-promote themselves into errors.  I'd prefer something like what @roystgnr suggested - where a flag must be specified in order to make it into an error.

**Comment by dschwen** [comment]
Well, that's not going to motivate MOOSE developers to follow through with their deprecations. Unless we always set the flag in testing and bully/shame devs that forget to remove things they deprecated and thus _hold up PRs for everyone_!

**Comment by dschwen** [comment]
@YaqiWang you mean put a patch "in escrow" with Civet?! That would be really cool. It would however
- take Civet development to add such a capabilty (i.e. maybe a special _patch branch_ that Civet periodically rebases onto the current devel)
- require handling of merge conflicts that may appear between deprecating and expiration point (when doing the automatic rebase Civet could message the _patch branch_ owner demanding a fix)

**Comment by lindsayad** [comment]
> @rwcarlsen I agree with most of this, except the "powerless to fix the problem" part. That's exactly what my top comment addresses, a CMD line switch to turn those errors back into warnings to permit the user to run stuff. The deprecation error could mention this possiblilty explicitly.

If this option is available to the user, are the developers who deprecated the code even going to *feel* this automated deprecation enforcement? Is this a large improvement over the current warning we emit in terms of whether it's going to stir action around removal of the deprecated code? If I was a user trying to run a simulation, even if I wasn't in a hurry, I'd just re-run with the flag you just told me about and go on my merry way.

**Comment by dschwen** [comment]
> I'd just re-run with the flag you just told me about and go on my merry way.

I was hoping it'd at least make you feel dirty.

**Comment by YaqiWang** [comment]
> @YaqiWang you mean put a patch "in escrow" with Civet?! That would be really cool. It would however
> 
> * take Civet development to add such a capabilty (i.e. maybe a special _patch branch_ that Civet periodically rebases onto the current devel to keep it fresh, and then automatically merges at the expiration point)
> * require handling of merge conflicts that may appear between deprecating and expiration point (when doing the automatic rebase Civet could message the _patch branch_ owner demanding a fix)

Yes. The chance of having merge conflicts will be rare. In that case, civet can post the PR and let developers intervene. This will force us to fix input files in the repositories when adding deprecation, which is something we often forget including myself.

**Comment by aeslaughter** [comment]
> Well, that's not going to motivate MOOSE developers to follow through with their deprecation. Unless we always set the flag in testing and bully/shame devs that forget to remove things they deprecated and thus _hold up PRs for everyone_!

It is not just the MOOSE developers inability to follow through. We deprecate things to allow applications time to update code, but that doesn't happen often. Thus, the deprecation lingers. For example, the MeshModifiers syntax was deprecated for a year or two. I finally got the itch to remove, but still needed to go update many applications, it took many weeks of my time.

We all have more work than time, so updating deprecated code is usually last on the list. Take a look at the CIVET dashboard, most of the apps have a "deprecated" target that is orange. We can't remove deprecated stuff in MOOSE until those are green.

**Comment by lindsayad** [comment]
I agree. You'll notice that in @roystgnr's https://github.com/idaholab/moose/issues/18631#issuecomment-904160193 it was all about how *users* interact with deprecation. We do deprecation in the hopes that we can remove some burden from ourselves in the applications, yet as @aeslaughter this sometimes doesn't happen. @YaqiWang's post above, especially before he edited it, reflects the general sentiment that most of our users seem to have: when we deprecate things, it's simultaneously our responsibility to go update the deprecation. That's not how it works for any other project that I'm involved in. It's always the users responsibility to go update deprecation not the upstream developer. There is an attitude, not shared by all mind you (for instance @gambka is a hero at addressing deprecation), that MOOSE developers should do everything. And it's exhausting.

**Comment by dschwen** [comment]
We could make deprecation dates machine readable and have Civet issue a report compiling all deprecation that have already expired. I just noticed that we just do not care about deprecations at all in modules and (probably most) apps. 

That way we could have deprecation warnings parsed by civet without making them errors (neither after the deprecation time nor before it). Also we'd not have to put the burden on the users either.

**Comment by friedmud** [comment]
After many years of going back and forth on this... I think I agree with the others stating that we need to take a harder stance.

The big issue (as has been brought up) is the intertwining of testing.  i.e. even if we make this a time bomb that is promoted to an error... it will "randomly" bring _all_ development to a hault just because one application didn't update.

Therefore I propose that deprecations are Warnings by default... then, with a flag, can be promoted to errors.  That flag should be used in all standalone testing of applications (i.e. the tests that fire for BISON PRs / merges but NOT the BISON tests that fire for MOOSE/other PR/merges).  This will allow those apps to start failing without impacting the rest of the ecosystem.

**Comment by roystgnr** [comment]
That's clever.

We've always had `--disable-deprecated` be a configure option in libMesh, so that we can have the preprocessor remove deprecated APIs entirely and get many deprecated-code-usage errors in a faster and easier to correct setting, at compile-time rather than run-time ... but even if you wanted that too, there's no reason you can't mix that with a run-time option, is there?  You could even just reuse the same option - someone compiling against a libMesh without `LIBMESH_ENABLE_DEPRECATED` on is definitely asking for compile-time errors; someone compiling against a normal libMesh probably wants everything to compile with warnings at worst in general, but they (or our CI) might still want a command line flag to promote the oldest warnings to errors, selectively, for some applications rather than others.

**Comment by friedmud** [comment]
@roystgnr - not a bad idea to be able to promote them to compile errors... but I'm not 100% certain that all can be checked at compile time.

We already have at least two types of deprecation: deprecated input file syntax... and deprecated APIs... and we handle them in two different ways.  This would need some thought.

**Comment by roystgnr** [comment]
I'd guess 30-50% of libMesh deprecations can be checked at compile time; I don't think it's ever been 100%.  But even catching 30% of old API usage at compile time is better than 0%.

Input files are the worst.  We've never dealt with deprecating input file syntax in libMesh (our policy: go invent your own syntax), but I watched Paul have to deal with that in GRINS and even though the eventual design was so much better as to be worth it, the code during the transition period was a mess.

**Comment by friedmud** [comment]
We actually have a good system for deprecating input file syntax... it's really not bad (which is one of the reasons we can keep the old syntax around for so long... it doesn't bother anyone).
